extended-tracer/src/main/java/io/opentelemetry/contrib/tracer/Tracing.java

```diff
@@ -0,0 +1,386 @@
+    baggage.forEach(builder::put);
+    Context context = builder.build().storeInContext(Context.current());
+    try (Scope ignore = context.makeCurrent()) {
+      return callable.call();
+    } catch (Throwable e) {
+      sneakyThrow(e);
+      return null;
+    }
+  }
+
+  /**
+   * Run the given {@link Runnable} inside a server span.
+   *
+   * <p>The span context will be extracted from the <code>transport</code>, which you usually get
+   * from HTTP headers of the metadata of an event you're processing.
+   *
+   * <p>If an exception is thrown by the {@link Callable}, the span will be marked as error, and the
+   * exception will be recorded.
+   *
+   * @param transport the transport where to extract the span context from
+   * @param spanBuilder the {@link SpanBuilder} to use
+   * @param callable the {@link Callable} to call
+   * @param <T> the type of the result
+   * @return the result of the {@link Callable}
+   */
+  public <T> T traceServerSpan(
+      Map<String, String> transport, SpanBuilder spanBuilder, Callable<T> callable) {
+    return extractAndRun(SERVER, transport, spanBuilder, callable, this::setSpanError);
+  }
+
+  /**
+   * Run the given {@link Runnable} inside a server span.
+   *
+   * <p>The span context will be extracted from the <code>transport</code>, which you usually get
+   * from HTTP headers of the metadata of an event you're processing.
+   *
+   * <p>If an exception is thrown by the {@link Runnable}, the handleException consumer will be
+   * called, giving you the opportunity to handle the exception and span in a custom way, e.g. not
+   * marking the span as error.
+   *
+   * @param transport the transport where to extract the span context from
+   * @param spanBuilder the {@link SpanBuilder} to use
+   * @param callable the {@link Callable} to call
+   * @param handleException the consumer to call when an exception is thrown
+   * @param <T> the type of the result
+   * @return the result of the {@link Callable}
+   */
+  public <T> T traceServerSpan(
+      Map<String, String> transport,
+      SpanBuilder spanBuilder,
+      Callable<T> callable,
+      BiConsumer<Span, Throwable> handleException) {
+    return extractAndRun(SERVER, transport, spanBuilder, callable, handleException);
+  }
+
+  /**
+   * Run the given {@link Runnable} inside a server span.
+   *
+   * <p>The span context will be extracted from the <code>transport</code>, which you usually get
+   * from HTTP headers of the metadata of an event you're processing.
+   *
+   * <p>If an exception is thrown by the {@link Callable}, the span will be marked as error, and the
+   * exception will be recorded.
+   *
+   * @param transport the transport where to extract the span context from
+   * @param spanBuilder the {@link SpanBuilder} to use
+   * @param callable the {@link Callable} to call
+   * @param <T> the type of the result
+   * @return the result of the {@link Callable}
+   */
+  public <T> T traceConsumerSpan(
```

Comment:
are separate `traceConsumerSpan` / `traceServerSpan` methods worth it now that they accept spanBuilder and user can setSpanKind before passing that in?

Reply:
I would say yes - it's guiding the user to set the span kind.

I also can't find a good name otherwise - for making this method public:

```java
private <T, E extends Throwable> T extractAndRun(
      SpanKind spanKind,
      Map<String, String> transport,
      SpanBuilder spanBuilder,
      ThrowingSupplier<T, E> throwingSupplier,
      BiConsumer<Span, Throwable> handleException)
      throws E {
    try (Scope ignore = extractContext(transport).makeCurrent()) {
      return call(spanBuilder.setSpanKind(spanKind), throwingSupplier, handleException);
    }
  }
```